AsyncDisplayKit/ASVideoPlayerNode.mm

```diff
@@ -0,0 +1,729 @@
+
+      if (_delegateFlags.delegateScrubberThumbImage) {
+        UIImage *thumbImage = [_delegate videoPlayerNodeScrubberThumbImage:self];
+        [slider setThumbImage:thumbImage forState:UIControlStateNormal];
+      }
+
+
+      [slider addTarget:self action:@selector(beginSeek) forControlEvents:UIControlEventTouchDown];
+      [slider addTarget:self action:@selector(endSeek) forControlEvents:UIControlEventTouchUpInside|UIControlEventTouchUpOutside|UIControlEventTouchCancel];
+      [slider addTarget:self action:@selector(seekTimeDidChange:) forControlEvents:UIControlEventValueChanged];
+
+      return slider;
+    }];
+
+    _scrubberNode.flexShrink = YES;
+
+    [_cachedControls setObject:_scrubberNode forKey:@(ASVideoPlayerNodeControlTypeScrubber)];
+  }
+
+  [self addSubnode:_scrubberNode];
+}
+
+- (void)createControlFlexGrowSpacer
+{
+  if (_controlFlexGrowSpacerSpec == nil) {
+    _controlFlexGrowSpacerSpec = [[ASStackLayoutSpec alloc] init];
+    _controlFlexGrowSpacerSpec.flexGrow = YES;
+  }
+
+  [_cachedControls setObject:_controlFlexGrowSpacerSpec forKey:@(ASVideoPlayerNodeControlTypeFlexGrowSpacer)];
+}
+
+- (void)updateDurationTimeLabel
+{
+  NSString *formatedDuration = [self timeStringForCMTime:_duration forTimeLabelType:ASVideoPlayerNodeControlTypeDurationText];
+  _durationTextNode.attributedString = [self timeLabelAttributedStringForString:formatedDuration forControlType:ASVideoPlayerNodeControlTypeDurationText];
+}
+
+- (void)updateElapsedTimeLabel:(NSTimeInterval)seconds
+{
+  NSString *formatedDuration = [self timeStringForCMTime:CMTimeMakeWithSeconds( seconds, _videoNode.periodicTimeObserverTimescale ) forTimeLabelType:ASVideoPlayerNodeControlTypeElapsedText];
+  _elapsedTextNode.attributedString = [self timeLabelAttributedStringForString:formatedDuration forControlType:ASVideoPlayerNodeControlTypeElapsedText];
+}
+
+- (NSAttributedString*)timeLabelAttributedStringForString:(NSString*)string forControlType:(ASVideoPlayerNodeControlType)controlType
+{
+  NSDictionary *options;
+  if (_delegateFlags.delegateTimeLabelAttributes) {
+    options = [_delegate videoPlayerNodeTimeLabelAttributes:self timeLabelType:controlType];
+  } else {
+    options = @{
+                NSFontAttributeName : [UIFont systemFontOfSize:12.0],
+                NSForegroundColorAttributeName: _defaultControlsColor
+                };
+  }
+
+
+  NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:string attributes:options];
+
+  return attributedString;
+}
+
+#pragma mark - ASVideoNodeDelegate
+- (void)videoNode:(ASVideoNode *)videoNode willChangePlayerState:(ASVideoNodePlayerState)state toState:(ASVideoNodePlayerState)toState
+{
+  if (_delegateFlags.delegateVideoNodeWillChangeState) {
+    [_delegate videoPlayerNode:self willChangeVideoNodeState:state toVideoNodeState:toState];
+  }
+
+  if (toState == ASVideoNodePlayerStateReadyToPlay && _durationTextNode) {
+    _duration = _videoNode.currentItem.duration;
```

Comment:
My hunch is telling me we want this duration to be captured even without the existence of the durationTextNode.  Aka the client want to log the duration of the video, but the actual durationTextNode isn't needed on the UI.